probe-rs/src/debug/unit_info.rs

```diff
@@ -0,0 +1,1587 @@
+                                    "Error: No Attribute Value for DW_AT_type for variable {:?}",
+                                    child_variable.name
+                                )));
+                                }
+                            }
+                        }
+                        Err(error) => {
+                            child_variable.set_value(VariableValue::Error(format!(
+                                "Error: Failed to decode pointer reference: {:?}",
+                                error
+                            )));
+                        }
+                    }
+                }
+                gimli::DW_TAG_union_type => {
+                    if let Some(child_member_index) = child_variable.member_index {
+                        match &parent_variable.memory_location {
+                            VariableLocation::Address(address) => {
+                                // This is a member of an array type, and needs special handling.
+                                let (location, has_overflowed) = address.overflowing_add(
+                                    child_member_index as u32 * child_variable.byte_size as u32,
+                                );
+
+                                // TODO:
+
+                                if has_overflowed {
+                                    return Err(DebugError::Other(anyhow::anyhow!(
+                                        "Overflow calculating variable address"
+                                    )));
+                                } else {
+                                    child_variable.memory_location =
+                                        VariableLocation::Address(location);
+                                }
+                            }
+                            _other => {
+                                child_variable.memory_location = VariableLocation::Unavailable;
+                            }
+                        }
+                    }
+
+                    // Recursively process a child types.
+                    // TODO: The DWARF does not currently hold information that allows decoding of which UNION arm is instantiated, so we have to display all available.
+                    child_variable = self.process_tree(
+                        node,
+                        child_variable,
+                        core,
+                        stack_frame_registers,
+                        cache,
+                    )?;
+                    if child_variable.is_valid() && !cache.has_children(&child_variable)? {
+                        // Empty structs don't have values.
+                        child_variable.set_value(VariableValue::Valid(format!(
+                            "{:?}",
+                            child_variable.type_name.clone()
+                        )));
+                    }
+                }
+                gimli::DW_TAG_subroutine_type => {
```

Comment:
Ultimately all of those branches should go into separate functions for better readability, but this can be done in another PR for better traceability.